File: backend-java/src/main/java/com/backend/java/backendjava/controllers/CardController.java
package com.backend.java.backendjava.controllers;

import com.backend.java.backendjava.adapters.CardAdapter;
import com.backend.java.backendjava.entities.Card;
import com.backend.java.backendjava.dtos.requests.CardRequest;
import com.backend.java.backendjava.dtos.responses.CardResponse;
import com.backend.java.backendjava.infra.services.CardService;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/cards")
public class CardController {
    @Autowired
    private CardService cardService;

    @PostMapping
    public ResponseEntity<CardResponse> createCard(@RequestBody @Valid CardRequest request) {
        CardResponse response = cardService.createCard(request);
        return ResponseEntity.status(201).body(response);
    }

    @GetMapping
    public ResponseEntity<Page<CardResponse>> readCard(Pageable pageable) {
        Page<CardResponse> responses = cardService.readCard(pageable);
        return ResponseEntity.status(200).body(responses);
    }

    @GetMapping("/{id}")
    public ResponseEntity<CardResponse> readCardById(Long id) {
        CardResponse response = cardService.readCardById(id);
        return ResponseEntity.status(200).body(response);
    }

    @PutMapping("/{id}")
    public ResponseEntity<CardResponse> updateCardById(@PathVariable Long id, @RequestBody @Valid CardRequest request) {
        CardResponse response = cardService.updateCardById(id, request);
        return ResponseEntity.status(200).body(response);
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteCardById(@PathVariable Long id) {
        cardService.deleteCardById(id);
        return ResponseEntity.status(204).body(null);
    }
}

File: backend-kotlin/src/main/kotlin/com/backend/kotlin/backendkotlin/controllers/CardController.kt
package com.backend.kotlin.backendkotlin.controllers

import com.backend.kotlin.backendkotlin.dtos.requests.CardRequest
import com.backend.kotlin.backendkotlin.dtos.responses.CardResponse
import com.backend.kotlin.backendkotlin.infra.services.CardService
import jakarta.validation.Valid
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*


@RestController
@RequestMapping("/api/cards")
class CardController {
    @Autowired
    private lateinit var cardService: CardService

    @PostMapping
    fun createCard(@RequestBody  @Valid request: CardRequest): ResponseEntity<CardResponse> {
        val response = cardService.createCard(request)
        return ResponseEntity.status(201).body(response)
    }

    @GetMapping
    fun readCard(pageable: Pageable): ResponseEntity<Page<CardResponse?>> {
        val responses = cardService.readCard(pageable)
        return ResponseEntity.status(200).body(responses)
    }

    @GetMapping("/{id}")
    fun readCardById(id: Long): ResponseEntity<CardResponse> {
        val response = cardService.readCardById(id)
        return ResponseEntity.status(200).body(response)
    }

    @PutMapping("/{id}")
    fun updateCardById(
        @PathVariable id: Long,
        @RequestBody  @Valid request: CardRequest
    ): ResponseEntity<CardResponse> {
        val response = cardService.updateCardById(id, request)
        return ResponseEntity.status(200).body(response)
    }

    @DeleteMapping("/{id}")
    fun deleteCardById(@PathVariable id: Long): ResponseEntity<Void> {
        cardService.deleteCardById(id)
        return ResponseEntity.status(204).body(null)
    }
}